### DISCUSSION TITLE
Delete custom function.

### DISCUSSION THREAD

**Discussion by crawe1** [YetiForceCompany/YetiForceCRM#15584] [How to]
Hi,
I created custom function to delete records - action -> delete. I have a problem with query. When i try to exequte simle query "SELECT * FROM u_yf_igrn_inventory" in igrn module I see all row from table u_yf_igrn_inventory but without this actual crmid row. 
@vovpff maybe have any idea why ?

**Comment by vovpff** [comment]
> I created custom function to delete records - action -> delete.

What you mean for that?

**Reply by crawe1** [reply]
workflow taks "when delete record" run custom function.

**Comment by vovpff** [comment]
Try

```
SELECT crmid FROM u_yf_igrn_inventory
```

**Reply by crawe1** [reply]
ok i try write step by step.
1) get record ID from IGRN module - ok for example $recordId = 1000
2) run select to table u_yf_igrn_inventory but in request i have 0 row. 
3) when i run select without WHERE parameter i have all row but without $recordId = 1000

its Magical situation becouse when i run this same query idatabase by dbweaver, phpmyadmin i have result. This situation is only in this workflow action "where delete record" 

 function IGRNWorkflowDelete(Vtiger_Record_Model $recordModel)

    {
		$idrecord = $recordModel->getId();
		$sqligrnDeleteQuery1 = "
		SELECT *
	        FROM u_yf_igrn_inventory
                WHERE u_yf_igrn_inventory.crmid = $idrecord ";
		$sqldeleteIgrnEmptyQuery = \App\Db::getInstance()->createCommand("$sqligrnDeleteQuery1")->queryall();
    }

**Reply by vovpff** [reply]
result of your query is array with a lot of data if `$idrecord` is valid crmid

Your query with my hardcocded igrn id 1658 (valid igrn in my system) return next array:

```
Array
(
    [0] => Array
        (
            [id] => 1
            [crmid] => 1658
            [seq] => 1
            [name] => 265
            [qty] => 300.000
            [price] => 0.00000000
            [total] => 0.00000000
            [comment1] => Тестовый продукт. Не трогать!!!
            [unit] => 
            [ean] => 
            [subunit] => 
            [prbk] => 
            [prid] => 
            [prpn] => 
            [qtyall] => 
        )

bla bla bla

   [2] => Array
        (
            [id] => 3
            [crmid] => 1658
            [seq] => 3
            [name] => 301
            [qty] => 200.000
            [price] => 0.00000000
            [total] => 0.00000000
            [comment1] => Тестовый продукт. Не трогать!!!еееее
            [unit] => 
            [ean] => 
            [subunit] => 
            [prbk] => 
            [prid] => 
            [prpn] => 
            [qtyall] => 
        )
```

Result of query

```
SELECT * FROM `u_yf_igrn_inventory` WHERE `crmid` = 1658
```

 via phpMyAdmin returns 

![image](https://user-images.githubusercontent.com/8545791/147505151-934ae026-d297-4ee0-b56c-2a3cbe37e2bb.png)

**Reply by crawe1** [reply]
Ok i try to count row hehe in yeti result is 2 rows but in table is 3 rows. 
Where is problem ? its simple query 
every one system back result without row crmid=idrecordu
SELECT COUNT(*) FROM u_yf_igrn_inventory

**Reply by vovpff** [reply]
I replaced 2nd table row with bla bla bla in my code. There is no problem with query

**Comment by crawe1** [comment]
@vovpff  but this situation is only when i delete record by workfow task. 

my example:
		SELECT *
			FROM u_yf_igrn_inventoryarray(2) {
  [0]=>
  array(20) {
    ["id"]=>
    int(483)
    ["crmid"]=>
    int(12833)
    ["seq"]=>
    int(1)
    ["name"]=>
    int(7676)
    ["qty"]=>
    string(7) "121.000"
    ["price"]=>
    string(11) "21.00000000"
    ["total"]=>
    string(13) "2541.00000000"
    ["comment1"]=>
    string(0) ""
    ["unit"]=>
    string(4) "szt."
    ["ean"]=>
    string(0) ""
    ["subunit"]=>
    string(0) ""
    ["gross"]=>
    string(13) "2541.00000000"
    ["currency"]=>
    int(1)
    ["currencyparam"]=>
    string(176) "{"1":{"date":"2021-12-14","value":1,"conversion":1},"2":{"date":"2021-12-14","value":4.6436,"conversion":0.21535},"3":{"date":"2021-12-14","value":4.1088,"conversion":0.24338}}"
    ["taxmode"]=>
    int(0)
    ["tax"]=>
    string(10) "0.00000000"
    ["taxparam"]=>
    string(1) "0"
    ["totalplnprice"]=>
    string(13) "2541.00000000"
    ["plnprice"]=>
    string(11) "21.00000000"
    ["uniquenumber"]=>
    int(128337676)
  }
  [1]=>
  array(20) {
    ["id"]=>
    int(484)
    ["crmid"]=>
    int(12847)
    ["seq"]=>
    int(1)
    ["name"]=>
    int(7676)
    ["qty"]=>
    string(6) "50.000"
    ["price"]=>
    string(12) "100.00000000"
    ["total"]=>
    string(13) "5000.00000000"
    ["comment1"]=>
    string(0) ""
    ["unit"]=>
    string(4) "szt."
    ["ean"]=>
    string(0) ""
    ["subunit"]=>
    string(0) ""
    ["gross"]=>
    string(13) "5000.00000000"
    ["currency"]=>
    int(1)
    ["currencyparam"]=>
    string(176) "{"1":{"date":"2021-12-24","value":1,"conversion":1},"2":{"date":"2021-12-24","value":4.6252,"conversion":0.21621},"3":{"date":"2021-12-24","value":4.0785,"conversion":0.24519}}"
    ["taxmode"]=>
    int(0)
    ["tax"]=>
    string(10) "0.00000000"
    ["taxparam"]=>
    string(1) "0"
    ["totalplnprice"]=>
    string(13) "5000.00000000"
    ["plnprice"]=>
    string(12) "100.00000000"
    ["uniquenumber"]=>
    int(128477676)
  }
}
![image](https://user-images.githubusercontent.com/61950059/147505798-768aeced-c3a3-4ca3-8a1e-d84a204a17bd.png)
this is id record model from i try delete

**Reply by vovpff** [reply]
Ok. Hmmmm. Maybe recordModel->getId() is empty when workflow fires afted deletion. Check it!

**Reply by crawe1** [reply]
on table i have 3 row but they back me only 2 row without this row from system where record id = crmid ok i do static query where crmid=12831 and i try delete this record. 

![image](https://user-images.githubusercontent.com/61950059/147506260-2e3a059d-503f-4f3a-b45e-34636ec62c89.png)

CLICK  BUTTON :)

query:
		$sqligrnDeleteQuery1 = "
		SELECT *
			FROM u_yf_igrn_inventory
			WHERE u_yf_igrn_inventory.crmid=12831";

PAUSE

RESULT 
		SELECT *
			FROM u_yf_igrn_inventory
			WHERE u_yf_igrn_inventory.crmid=12831array(0) {
}

0 in array

now in table form database:
![image](https://user-images.githubusercontent.com/61950059/147506427-8bac831c-aa1e-4735-b148-62847dce1172.png)

this same id module id from module and id from query i not have result only in workflow task "where delete record"

**Reply by crawe1** [reply]
> 

Ok. Hmmmm. Maybe recordModel->getId() is empty when workflow fires afted deletion. Check it! <
hmm, no system back correct id this same as module id.

**Reply by vovpff** [reply]
I don't have ideas more. 

Try to implement the same process via handler.

**Reply by crawe1** [reply]
i hardcoded id in query but not result 
SELECT *
FROM u_yf_igrn_inventory
WHERE u_yf_igrn_inventory.crmid=12831array(0)

but this stituation is 
1) where i delete record
2) where deleted module id is this same as crmid if i change situation on static  that delete modeule id != crmid everything is ok

**Comment by crawe1** [comment]
@vovpff Voila ! befeoreDelete  handler resolve problem thankyou for good help!

**Reply by vovpff** [reply]
My congratulations!

**Comment by crawe1** [comment]
@vovpff @mariuszkrzaczkowski 
I have one more question tell me if it possible add return in this hendler beforeDelete in function "entityBeforeDelete" in such a way that stop execute and result message with error ? How it stop with message ? this result work form me in hendler "preSave" handler but in this  hendler not work.

return $response = [**'result' => false**, 'message' => App\Language::translate($outputText, $recordModel->getModuleName())];

**Reply by mariuszkrzaczkowski** [reply]
You can only add an error exception
